Simplify code product alert email

arnoudhgz opened the pull request:
The send method was complex and had some duplication in code. This is improved by using private methods for small parts of this method and making the if/else statements more logical (else isn't even used anymore).

As result the PHPMD suppress warning annotations could be removed.

For better string comparison the currently defined types (price and stock) are set as constants.

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Reduce the complexity of the send method

The send method was complex and had some duplication in code. This is
improved by using private methods for small parts of this method and
making the if/else statements more logical (else isn't even used
anymore).

As result the PHPMD suppress warning annotations could be removed.

magento-engcom-team commented:
Hi @arnoudhgz. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me {$VERSION} instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://github.com/magento/magento2/wiki/Magento-Contributor-Assistant) documentation

orlangur commented on app/code/Magento/ProductAlert/Model/Email.php:
```diff
@@ -286,110 +319,136 @@ protected function _getStockBlock()
+            return self::XML_PATH_EMAIL_PRICE_TEMPLATE;
+        }
+
+        return self::XML_PATH_EMAIL_STOCK_TEMPLATE;
```
Please replace all such conditionals with ternary operator.

arnoudhgz replied:
@orlangur thank you for your feedback, I changed it...

VladimirZaets commented:
Hi @arnoudhgz, thank you for collaboration. Please enable `strict type` mode for this uses `declare(strict_types=1)` declaration.

VladimirZaets commented on app/code/Magento/ProductAlert/Model/Email.php:
```diff
@@ -286,110 +319,128 @@ protected function _getStockBlock()
+     *
+     * @return StoreInterface|Store
+     */
+    private function getStore($customerStoreId)
```
Please use `type hints` for the method arguments.

VladimirZaets commented on app/code/Magento/ProductAlert/Model/Email.php:
```diff
@@ -286,110 +319,128 @@ protected function _getStockBlock()
+     *
+     * @return string
+     */
+    private function getTemplateConfigPath()
```
Please use `type hints` for the method arguments.

VladimirZaets commented on app/code/Magento/ProductAlert/Model/Email.php:
```diff
@@ -286,110 +319,128 @@ protected function _getStockBlock()
+     *
+     * @return array
+     */
+    private function getProducts()
```
Please use `type hints` for the method arguments.

VladimirZaets commented on app/code/Magento/ProductAlert/Model/Email.php:
```diff
@@ -286,110 +319,128 @@ protected function _getStockBlock()
+     *
+     * @return Price|Stock
+     */
+    private function getBlock()
```
Please use `type hints` for the method arguments.

arnoudhgz commented:
@VladimirZaets  I have added the strict types and included type hinting where possible.

I saw some existing protected methods like `setWebsiteId` where I could not define a strict type, because the input can be also string in dependend code....

arnoudhgz commented:
@VladimirZaets I see that with the strict types give now errors in the unit tests. I will first fix this before you have to check again.

arnoudhgz commented:
@VladimirZaets  I will fix the errors in the unit test first before you have to check again.

VladimirZaets commented:
Hi @arnoudhgz, maybe you need any help?

arnoudhgz commented:
@VladimirZaets  Yes,  I would like that. 
I am struggling with the strict type for the `setCustomerData` method. First I had `CustomerInterface` but that did not workout, `DataObject` worked for the unit tests, but now it fails on the integration test.

VladimirZaets commented:
@arnoudhgz as I see, the `$customer` should be `CustomerInterface` type. I searched usages of this method but I can't find any place where was passed param with other type. Can you please explain what the problem when you add `CustomerInterface` as a type?

arnoudhgz commented:
@VladimirZaets  I had this first as `CustomerInterface` but than the unit tests said that a `DataObject` was given. The `CustomerRepositoryInterface` has a return type of `CustomerInterface` but it actually returns a `DataObject`. 

That is why I changed it to `DataObject`.

VladimirZaets commented:
@arnaucoll I think the problem is in the unit test that returns DataObject. What test is failed?

arnoudhgz commented:
@VladimirZaets  please see my updated comment with screenshot ^^

VladimirZaets commented:
@arnoudhgz The problem is in the test. 
In [this](https://github.com/magento/magento2/blob/6b7a9b3d3529ea9518fab12d0852afb3c73dc649/app/code/Magento/ProductAlert/Test/Unit/Model/ObserverTest.php#L299) line uses `DataObject` instead of `CustomerInterface` to create the customer mock. 

Can you please fix it?

arnoudhgz commented:
@VladimirZaets it seems solved now. There is only the timeout in Travis.

VladimirZaets commented:
@arnoudhgz Thanks for fixes

VladimirZaets commented:
Hi @arnoudhgz, can you please make refactoring and use private properties instead of constants.

arnoudhgz commented:
@VladimirZaets I can, but it seems wrong, I think the class `Magento\ProductAlert\Model\Observer` should use `EMAIL::TYPE_PRICE` abd `EMAIL::TYPE_STOCK` instead of hardcoded strings in `$email->setType('price');` and `$email->setType('stock');`.

Do you think I should maybe refactor this Observer also to make the code better (remove complexities)?

arnoudhgz commented:
@VladimirZaets are you really sure I should use private properties? These values are constants imho, they can't be changed with setters or inside other functions. 

According to the defnition of constants on the PHP website:
> A constant is an identifier (name) for a simple value. As the name suggests, that value cannot change during the execution of the script

VladimirZaets commented:
Hi @arnoudhgz. I discussed this question and we take decisions what we should do in the similar cases. 
We should use constants if value uses in the few classes. If their value uses in the scope of the class we should use private static declaration.

arnoudhgz commented:
@VladimirZaets thank you for the explanation, after I changed it to private static variables some test failed in Travis. It was about having to many variables on the class (the maximum is 15). So I changed the code to compare on strings (price or stock).

VladimirZaets commented:
Hi @arnoudhgz. I discussed this issue with my teammates and we take the decision that we can't add strict type and type checking to the already existing files due to [backward-compatible reason](http://devdocs.magento.com/guides/v2.2/contributor-guide/backward-compatible-development/). It's my fail, sorry. Can you please revert the changes related to strict types. Thanks

arnoudhgz commented:
@VladimirZaets  I will change it back.... :+1:

arnoudhgz commented:
@VladimirZaets I removed the strict types checking + the return types and parameter typehinting for existing functions.

arnoudhgz commented:
@VladimirZaets  do I still need to update the code? Could you provide some feedback?

VladimirZaets commented:
Hi @arnoudhgz. Sorry for the delay, we took your PR for delivery

VladimirZaets committed:
Simplify code product alert email

magento-engcom-team commented:
Hi @arnoudhgz. Thank you for your contribution. 
We will aim to release these changes as part of 2.3.0. 
Please check the release notes for final confirmation.
